### ISSUE TITLE
v4.0.0-RC branch takes forever to compile

### ISSUE THREAD

**Issue by JensMadsen** [boutproject/BOUT-dev#382]
with 

  $ ./configure --enable-debug --with-fftw=/media/xtradisk/numerics/fftw --enable-track --with-netcdf=/home/media/xtradisk/numerics/netcdf-4.4.1

make clean 
time make 

yields

real	13m22.303s
user	2m38.232s
sys	10m14.340s

On master with 

./configure --with-debug --with-fftw=/media/xtradisk/numerics/fftw --with-tr
ack

make clean 
time make

yields: 

real	1m27.921s
user	0m54.052s
sys	0m13.228s
 
so next branch take 8-9 times longer to compile?

**Comment by dschwoerer** [comment]
It seems a huge amount in compiling next is spent in sys, unlike in master.
Did you make sure to run `make clean; time make` twice, to make sure the files are in cache?

I would assume that it takes longer, as more functions are defined in headers or using templates.
That is needed to allow inlining, allowing better performance at run time - at the cost of compile time ...
Looking at user - the it only takes 2 - 3 times as long.

I compiled also master and next, and the difference is also factor 2-3.

Using /usr/bin/time instead of the built-in, shows you the major page faults
https://linux.die.net/man/1/time

Finally, running `make clean; make -j 8;make` should speed things up ...

**Comment by ZedThree** [comment]
This is a little surprising. Can you post the `BOUT_FLAGS` line from `make.config`?

**Comment by JensMadsen** [comment]
@dschwoerer I am not sure that I understand your comment about that the files are in the cache? 

I tried to compile again and I got : 

real	15m8.742s
user	2m42.732s
sys	10m6.616s

@ZedThree 

BOUT_FLAGS		= $(CXXFLAGS) -g -O2 -g -O0 -Og -DCHECK=3 -DSIGHANDLE -DTRACK -DBACKTRACE -DNCDF -DBOUT_HAS_PVODE  -std=c++11

**Comment by d7919** [comment]
Can you try removing all the `-g` entries and see if that improves matters? I've found this can potentially make a large difference.

**Comment by dschwoerer** [comment]
The OS tries to keep the files in RAM, but it seems in your case, they are not.
Can you try::
`make clean;make;make clean; time make`
Again, you have 10 minutes in sys calls, which is most likley reading files from disk, and only 2:42 actually working.

**Comment by JensMadsen** [comment]
Tried your suggestion @dschwoerer 

real	6m21.525s
user	2m38.264s
sys	3m19.464s

still more than 4 times longer than master, but better:-D

**Comment by dschwoerer** [comment]
That's unexpected. Can you use /usr/bin/time?
`make clean;make;make clean; /usr/bin/time make`

**Comment by JensMadsen** [comment]
yes : 

real	6m9.525s
user	2m39.656s
sys	3m0.476s

**Comment by dschwoerer** [comment]
I thought `/usr/bin/time` would be GNU's time ...
I would be keen to know how many major page fauts there are - as I assume that that is causing the high sys time.

**Comment by JensMadsen** [comment]
I just ran top in another terminal during compilation. ar is on 100 % most of the time? What does that mean?

**Comment by loeiten** [comment]
I also have this problem.
Timing the building time in `next` gives

```
real	8m37.742s
user	7m40.480s
sys	0m45.708s
```

It looked like @JensMadsen problem ceased on his new computer, but it would be a pity if the only thing which could save long build-time would be a new computer. Is there a different solution to this problem?

**Comment by d7919** [comment]
I find removing `-g` often helps speed up the compilation (actually the creation of the library) quite a lot -- obviously not that helpful if you actually want the symbols but may sometimes be useful. 

Another option that might help is to delaying the library creation until after everything has been compiled.

```
lib:  ........
ifneq ("$(OBJ)foo", "foo")
    @echo "Adding ......"
    @$(AR).......
    @$(RANLIB).......
endif

else
:::

```

with

```
lib:  ........

else

arlib:
    @echo "Adding ......"
    @$(AR) $(ARFLAGS) $(LIB) $(shell find ./src -name \*.o)
    @$(RANLIB).......
```

You can then do `make -j && make arlib` , the first make command just compiles all the source files (in parallel) whilst the second combines all the object files into the library. This may or may not help.

**Comment by dschwoerer** [comment]
Old:
```
$ make clean &> /dev/null; time ( make &> /dev/null) ; time ( make &> /dev/null)
real    3m35.182s
user    1m51.742s
sys     0m13.616s

real    0m46.209s
user    0m2.404s
sys     0m5.868s
```
with #531 (related to @d7919 proposal)
```
$ make clean &> /dev/null; time ( make &> /dev/null) ; time ( make &> /dev/null)
real    2m43.386s
user    1m49.910s
sys     0m8.061s

real    0m4.947s
user    0m1.439s
sys     0m0.328s
```

**Comment by dschwoerer** [comment]
And parallel:
```
real    0m42.106s
user    1m59.209s
sys     0m11.756s

real    0m10.416s
user    0m1.678s
sys     0m1.669s

```

```
$ make clean &> /dev/null; time ( make -j8 &> /dev/null) ; time ( make -j8 &> /dev/null)
real    0m37.291s
user    1m57.308s
sys     0m8.374s

real    0m2.446s
user    0m1.397s
sys     0m0.298s
```
surprising little difference :/
however, #531 produces a complete library, even if build in parallel. So the additional `make` isn't needed (which therefore saves ~46 s and typing.

**Comment by bendudson** [comment]
I think this has now been fixed by #531 , or at least is about as good as it's going to get.
